File: LeetCodeCPlusPlus/py23.cpp

#include <iostream>
#include <queue>
#include <vector>
using namespace std;
// 合并K个有序链表，优先队列实现
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int,int> > > queue; // 注意>>代表右移，所以空格不能少
        // greater对应小顶堆、less默认对应大顶堆
        for (int i=0;i<lists.size();i++)
            if (lists[i]!=NULL) {
                queue.push(pair<int,int>(lists[i]->val, i));
                lists[i] = lists[i]->next;
            }
        ListNode* result = new ListNode(0);
        ListNode* p = result;
        while(!queue.empty()) {
            pair<int,int>tmp = queue.top();  // 获取队头
            p->next = new ListNode(tmp.first);
            p = p->next;
            queue.pop();  // 弹出
            if (lists[tmp.second]!=NULL) {
                queue.push(pair<int,int>(lists[tmp.second]->val, tmp.second)); // 不空，继续加入队
                lists[tmp.second] = lists[tmp.second]->next;
            }
        }
        return result->next;      
        
    }
};

File: LeetCodePython/py23.py
from typing import List 
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# 合并K个有序链表
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        # 来源于LeetCode-cn用户powcai, 堆排序优化
        import heapq
        dummy = ListNode(0)
        p = dummy
        head = []
        for i in range(len(lists)):  # 将每一个链第一个节点放入堆中
            if lists[i] :
                heapq.heappush(head, (lists[i].val, i))
                lists[i] = lists[i].next
        while head:
            val, idx = heapq.heappop(head)  # 弹出，新建
            p.next = ListNode(val)
            p = p.next
            if lists[idx]:  # 不为空 
                heapq.heappush(head, (lists[idx].val, idx))  # 放入下一个 
                lists[idx] = lists[idx].next
        return dummy.next

class Solution2:
    # 分治
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists:
            return []
        if len(lists) == 1:
            return lists[0]
        elif len(lists) == 2:
            result = ListNode(0)
            p = result
            first = lists[0]
            second = lists[1]
            while first and second:
                if first.val<=second.val:
                    p.next = first
                    p = p.next
                    first = first.next
                else:
                    p.next = second
                    p = p.next
                    second = second.next
            if first:
                p.next = first
            if second:
                p.next = second
            return result.next
        else:
            left = self.mergeKLists(lists[:len(lists)//2])  
            right = self.mergeKLists(lists[len(lists)//2:])  
            return self.mergeKLists([left,right])     



        
        